### ISSUE TITLE
Cannot accept value that would put our exposure to dust HTLCs

### ISSUE THREAD

**Issue by holzeis** [get10101/10101#580]
Our users currently can't fund their 10101 wallet because of that error.

```
2023-05-03 09:50:28 | {"timestamp":"2023-05-03T09:50:28.484712603Z","level":"INFO","fields":{"message":"Cannot accept value that would put our exposure to dust HTLCs at 10201000 over the limit 5000000 on counterparty commitment tx","log.target":"lightning::ln::channel","log.module_path":"lightning::ln::channel","log.file":"/home/runner/.cargo/git/checkouts/rust-lightning-8ffd0e8d14b5d2cc/6df5761/lightning/src/ln/channel.rs","log.line":2842},"target":"lightning::ln::channel"}
-- | --
2023-05-03 09:50:28 | {"timestamp":"2023-05-03T09:50:28.484820391Z","level":"INFO","fields":{"message":"Cannot accept value that would put our exposure to dust HTLCs at 10201000 over the limit 5000000 on holder commitment tx","log.target":"lightning::ln::channel","log.module_path":"lightning::ln::channel","log.file":"/home/runner/.cargo/git/checkouts/rust-lightning-8ffd0e8d14b5d2cc/6df5761/lightning/src/ln/channel.rs","log.line":2852},"target":"lightning::ln::channel"}
```

**Comment by holzeis** [comment]
It looks like this issue is due to some unreasonable high transaction fees at the moment. I guess whats happening is that the `htlc_timeout_dust_limit` and `htlc_success_dust_limit` are too high for a small payment of 15_000. (that was the biggest payment I tried) and our `max_dust_htlc_exposure_msat` is set to 5_000 sats which is smaller than 15_000.

![Screenshot 2023-05-03 at 12 54 41](https://user-images.githubusercontent.com/382048/235897062-b4150fc5-2d5a-4074-a94c-3315b86574c4.png)

https://github.com/get10101/rust-lightning/blob/6df5761cfc4a27c35ca26a547ff90f656d50d85f/lightning/src/ln/channel.rs#L2831-L2856

```
		let (htlc_timeout_dust_limit, htlc_success_dust_limit) = if self.opt_anchors() {
			(0, 0)
		} else {
			let dust_buffer_feerate = self.get_dust_buffer_feerate(None) as u64;
			(dust_buffer_feerate * htlc_timeout_tx_weight(false) / 1000,
				dust_buffer_feerate * htlc_success_tx_weight(false) / 1000)
		};
		let exposure_dust_limit_timeout_sats = htlc_timeout_dust_limit + self.counterparty_dust_limit_satoshis;
		if msg.amount_msat / 1000 < exposure_dust_limit_timeout_sats {
			let on_counterparty_tx_dust_htlc_exposure_msat = inbound_stats.on_counterparty_tx_dust_exposure_msat + outbound_stats.on_counterparty_tx_dust_exposure_msat + msg.amount_msat;
			if on_counterparty_tx_dust_htlc_exposure_msat > self.get_max_dust_htlc_exposure_msat() {
				log_info!(logger, "Cannot accept value that would put our exposure to dust HTLCs at {} over the limit {} on counterparty commitment tx",
					on_counterparty_tx_dust_htlc_exposure_msat, self.get_max_dust_htlc_exposure_msat());
				pending_forward_status = create_pending_htlc_status(self, pending_forward_status, 0x1000|7);
			}
		}


		let exposure_dust_limit_success_sats = htlc_success_dust_limit + self.holder_dust_limit_satoshis;
		if msg.amount_msat / 1000 < exposure_dust_limit_success_sats {
			let on_holder_tx_dust_htlc_exposure_msat = inbound_stats.on_holder_tx_dust_exposure_msat + outbound_stats.on_holder_tx_dust_exposure_msat + msg.amount_msat;
			if on_holder_tx_dust_htlc_exposure_msat > self.get_max_dust_htlc_exposure_msat() {
				log_info!(logger, "Cannot accept value that would put our exposure to dust HTLCs at {} over the limit {} on holder commitment tx",
					on_holder_tx_dust_htlc_exposure_msat, self.get_max_dust_htlc_exposure_msat());
				pending_forward_status = create_pending_htlc_status(self, pending_forward_status, 0x1000|7);
			}
		}
```

**Comment by holzeis** [comment]
Tested the following payments.

- 15k sats: failed
- 20k sats: failed
- 22k sats: failed
- 25k sats: failed
- 30k sats: success

**Comment by holzeis** [comment]
Use the following calculation to approximate the amount of sats that you can send to the 10101 wallet.

1. Fetch the current fees from mempool
```
 curl -sSL "https://mempool.space/api/v1/fees/recommended"
{"fastestFee":48,"halfHourFee":34,"hourFee":27,"economyFee":8,"minimumFee":4}
```
2. We are estimating our fees with `Normal` targeting for 3 blocks. I couldn't find an public api that would give me the 3 blocks target estimated fee directly, so we have to go with the hourFee as closest estimate here.
3. Calculate acceptable amount.

```
253 * {hourFee} * 1250 / 1000 * 663 / 1000
```

**Comment by holzeis** [comment]
@bonomat pointed out one solution to this issue would be to make use of [AnchorOutputs](https://bitcoinops.org/en/topics/anchor-outputs/). For that we have to update our ldk fork to the latest upstream.

**Comment by bonomat** [comment]
Strictly speaking: this not a bug but by design 😅🙈

**Comment by bonomat** [comment]
> Strictly speaking: this not a bug but by design 😅🙈

Added bug label again. I think it should still work 😅

**Comment by bonomat** [comment]
> > Strictly speaking: this not a bug but by design 😅🙈
> 
> Added bug label again. I think it should still work 😅

Actually, maybe not? 🤔

**Comment by holzeis** [comment]
> > > Strictly speaking: this not a bug but by design 😅🙈
> > 
> > 
> > Added bug label again. I think it should still work 😅
> 
> Actually, maybe not? 🤔 Is this ticket even relevant then?

You lost me 😅 

We can add another ticket for adding Anchor Outputs support, if you don't want to track it on this ticket.

**Comment by bonomat** [comment]
> You lost me 😅

Not just your 🙈

> We can add another ticket for adding Anchor Outputs support, if you don't want to track it on this ticket.

Ah, ok, so this ticket gets resolved once we upgrade to anchor outputs. Let's keep it then :)

**Comment by holzeis** [comment]
@bonomat is this still MVP relevant, given that we are most likely fixing the issue with removing the upper limit of tx fees?

If not, I would close that ticket and open another one for introducing AnchorOutputs.

**Comment by bonomat** [comment]
I think it is an issue until we have anchor outputs: when adding HTLCs we have to account for the TX fee and if we can't use CPFP to bump the fee, the reserved fee amount might render the remaining amount below dust limit. That's what makes HTLCs fail

**Comment by holzeis** [comment]
blocked by #589

**Comment by klochowicz** [comment]
We settled on temporarily increasing the minimum required invoice size to 50k, until we have proper anchor output support added in #661

**Comment by holzeis** [comment]
Increasing the invoice size to 50k does not solve that issue. Reopening, for the anchor output fix.

**Comment by klochowicz** [comment]
I rebased the draft PR. However, after reading a bit into LDK discord and roadmap, I concluded that we can't do much about it until anchor outputs land and LDK (and our DLC fork is updated afterwards.).
At this stage there is *hope* they will land in upcoming 116 release (by the end of Q2). I subscribed to relevant threads, and will update the ticket.

In the meantime, marking it as 'blocked'
